### ISSUE TITLE
Add CLI call feature for overriding any config.json parameter

### ISSUE THREAD

**Issue by mpompolas** [ivadomed/ivadomed#941]
Following the latest lab meeting discussion, we are considering adding a feature so we can call the Ivadomed framework in the command line, by overriding any parameter from the config.json file. 
CLI parameter override already exists in ivadomed but only for a few parameter (namely --train/test/segment, --path-data, --path-output etc.)

As @charleygros mentioned during the meeting, the proposed feature is already implemented within the SCT toolbox, so we can get inspired for this functionality from sct_run_batch

**Comment by dyt811** [comment]
Hey, @mpompolas , So I reviewed the issue. If I am understanding, the basic requirement of the features are:

- There should be a Model/Singleton where all default key/parameters are defined with default/values/settings preconfigured (currently this is being take care  of by `config.json` as a default template). 
- Work with a model configuration that allow override from CLI arguments. 
- Ideally and optionally, work with a model configuration that allow override from ANOTHER specification/configuration file (JSON/YAML/ENV etc)
- Ideally and optionally, the CLI argument key will be auto checked against the "model" to ensure it is anticipated. 
- Ideally and optionally, the CLI argument key data type will be checked and validated before the actual CLI commands move forwards 

- Current Model  (Status Quo)
   * JSON based configuration as a MANDATORY field for pretty much all ivadomed binaries
   * Have a default JSON (that user needs to duplicate and update)
   * CLI parameters are not generatelly from JSON. 
   * CLI parameters are not staticically enforced or centrally managed .
   * Each `/script/` binaries does its own argument parsing, for specific purposes per command SOME of these (but not all) are overwriting specific portion of the configuration JSON

So I dug this a bit more, it is a complex topics with many choices than I thought possible. 

This is a good [introductory article](https://towardsdatascience.com/from-novice-to-expert-how-to-write-a-configuration-file-in-python-273e171a8eb3) on the complexity of the issue. 

- YACS: 
   * it is very MODEL driven. A prototype nodal based default model is defined. Then, it has built in functions to be used to MERGE from file or gets specific parameters via cli 
   * Most notably, it requires YAML configuration migration away from the current JSON approach
   * still require manually CLI parameter parsing, handling/typechecking

- Dynaconf
   * Support `JSON`, `YAML`, `ini`, `env` etc and can load everything into a EASILY accessible object/class/dict. 
   * Downside is CLI arguments still require manual handling, parsing. 

- OmegaConf or Facebook Hydra: 
   * Require migration to YAML/YML. Probably allow the most sophisticated multi-tier YAML overwrite 
   * Very strong type checking
   * Downside is CLI arguments still require manual handling, parsing. 

- Confuse: 
   * Very good YAML layering and also CLI generation with validation 
   * Does not do JSON configuration file. 

- Typer/Click/Google Fire:
   * More about CLI argument (auto)generation
   * No JSON/YAML/YML support. Usually are about generate Python function arguments into  ready to use CLI parameters/arguments. 

- JSONargparse:
   * seems more about parsing JSON into Arguments for CLI where as our needs is to use CLI to overwrite prototypical/model config JSON

The problems I see are multifold:
* There so far does not appear to have JSON ready-to-use CLI generation tool out there that suits our needs. 
* JSON based configuring/modelling library does not appear to be popular. So far all config layering are done using YAML/YML files.
* [Since YAML is apparently a superset of JSON](https://en.wikipedia.org/wiki/JSON#YAML), we can technically migrate our entire under the hood configuration system to one of the above WHILE maintaining superficial Python configuration (and essentially adding YAML file based configuration support)
* IF we stick with JSON, We probably can only support first level SIMPLE key overwrite at CLI.  Since there are COMMON keywords across multiple first level key. 
![image](https://user-images.githubusercontent.com/6720068/133003159-7961936e-af1b-4f95-aa3e-329e32a7e80f.png)
e.g. We cannot accept e.g. `--applied false` blindly. So that either means 1) we manually generate the parameter for use at CLI level and map those back to the config.json level which is not really scalable or feasible as the system become more and more complex. or 2) we move towards a more modal based system and HOPEFULLY (since none of the above auto validate CLI level parameters passed to it) we have a way to traverse all the keys and validate the key overwritten is correct. 

Possible Path Going Forward
1. If we decide to Stick with JSON, probably DynaConf is the way for unified configuration.  We will code the parameter/proper check at CLI level if we want to offer configure eveyrthing at CLI  level. Some sort of scheme needs to be generated (e.g. `--modified3dunet.applied true`. JSON overwriting/layering is currently non-existent, if my understanding of the code base is correct. 
2. If we decide to add the JSON to YAML transpiling layer and migrate the underlying system to YAML completely and gradually phase users off JSON towards YAML, it  can add some interesting benefits (such as more data type support, less double quote related issues. You can find more pro vs con [here](https://octopus.com/blog/state-of-config-file-formats). 
3. We can just add the `GPU ID` in the current arg.parser and kick the can down the road until we really hear user demanding CLI overwrite of parameters and add/adapt those as needed. 

So in summary:
* Most libraries (except DynaConf) uses YAML to do configuration layering/combination
* So far, have not discovered any auto CLI parameters generation process that leverage JSON spec/config file. 
* We are currently manually managing at CLI/Script level what parameters gets overwritten. Entirely manually managed at PER cli command binary level. No centralization. 
* Choices are: Stay JSON, hoodwink YAML or only add optional parameters when people ask for it.
